docs: dan v2 overview (#38)
tari-project/rfcs: src/RFC-0303_DanOverview.md

THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

## Language

The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", 
"NOT RECOMMENDED", "MAY" and "OPTIONAL" in this document are to be interpreted as described in 
[BCP 14](https://tools.ietf.org/html/bcp14) (covering RFC2119 and RFC8174) when, and only when, they appear in all capitals, as 
shown here.

## Disclaimer

This document and its content are intended for information purposes only and may be subject to change or update
without notice.

This document may include preliminary concepts that may or may not be in the process of being developed by the Tari
community. The release of this document is intended solely for review and discussion by the community of the
technological merits of the potential system outlined herein.

## Goals

The aim of this Request for Comment (RFC) is to describe the key elements of the Tari second layer, also known as the 
Digital Assets Network (DAN).

## Related Requests for Comment

* [RFC-0100: Base Layer](RFC-0100_BaseLayer.md)

## Description

> The Tari DAN utilises a sharded BFT consensus mechanism called [Cerberus](https://arxiv.org/abs/2008.04450). 

Reviewer: There may be some misinterpretations about which parts of Cerberus are used
```suggestion
The Tari DAN is based on a sharded BFT consensus mechanism called [Cerberus](https://arxiv.org/abs/2008.04450). 
```

Reviewer: Of particular note is that the paper uses pBFT (3 rounds) whereas Tari currently uses Hotstuff (4 rounds)

registration deposit also serves as a Sybil prevention mechanism.

Validator nodes are the bridge between the consensus layer and the Tari Virtual Machine (TVM).

VNs are required to re-register periodically as a proof-of-liveness mechanism.

Validator nodes must be able to
1. interpret the instructions they receive from clients, identifying the contract code that the instruction refers,
2. retrieve and deserialize the relevant input state,
3. compute the output state that result from applying the contract logic to the input state, and
4. reach consensus with its peers.

Steps 1 - 3 are carried out in the Tari Virtual Machine (TVM).
Step 4 is achieved by communicating with peers via the DAN consensus layer.

[Thaums](#thaums-and-the-turbine-model) exist at the Validator node level, and VNs earn fees, in Thaums, for each 
instruction that it aids in getting finalised.

### DAN consensus layer

The Cerberus BFT consensus algorithm runs on the consensus layer. This layer is completely ignorant of the 
semantics of DAN smart contracts.

This layer only cares that:
* _only_ VNs that have registered on the base layer are participating in consensus.

> * VNs are self-organising into VN committees and are carrying out the rules of Cerberus and emergent HotStuff 

Reviewer: Emergent Hotstuff has not be defined or reference anywhere. I think its inclusion here is unneeded
```suggestion
* VNs are self-organising into VN committees and are carrying out the rules of Cerberus
```

  correctly.

In particular, the consensus layer has _no idea_ whether an instruction's output is correct. If two-thirds (plus one)

> of the committee say that the results are correct, then consensus has been reached and the consensus layer is happy.

Reviewer: ```suggestion
of the committee agree on the results, then consensus has been reached and the consensus layer is happy.
```

Reviewer: It's perhaps also worth mentioning that the result they agree on could be an error in executing the code.

implementation for TVM contracts, but in principle, other languages could implement the specification as well.

The TVM is able to
* load a contract.
* provide a list of methods that the contract exposes.
* Execute calls on the contract.
* Persist and restore the state of the contract.

### Thaums and the turbine model

Thaums is a [unit of magic](https://discworld.fandom.com/wiki/Thaum). Thaums are used to power the DAN's economic 
engine. Thaums are created with a one-way perpetual peg as described in [RFC-0320]. Briefly, Tari is burnt to create 
Thaums, which are used to pay for the execution of instructions on the DAN. A portion of instruction fees are burnt 
with every instruction to provide a constant source of demand for Thaums in the DAN.

There is no peg out back to the base layer for Thaums. Thaum holders wishing to convert back to Tari will be able to 
perform a submarine swap with a Tari seller. We also anticipate that exchanges will list the Thaum-Tari pair to 
enable easy conversion of Thaums back to Tari.

# Change Log

| Date        | Change              | Author     |
|:------------|:--------------------|:-----------|

> | 1 Nov 2022  | High-level overview | CjS77      |

Reviewer: Some RFCs have this in ascending order, some in descending. Not sure what the standard should be

Author: I also noticed the lack of consistency and decided to go with it for now :)
